docs/man/memcached_callback.3
.TH "MEMCACHED_CALLBACK" "3" "September 16, 2012" "1.0.10" "libmemcached"
.SH NAME
memcached_callback \- libmemcached Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" Man page generated from reStructeredText.
.
.sp
Get and set a callback
.SH SYNOPSIS
.sp
#include <libmemcached/memcached.h>
.INDENT 0.0
.TP
.B memcached_callback_t
.UNINDENT
.INDENT 0.0
.TP
.B memcached_return_t memcached_callback_set(memcached_st\fI\ *ptr\fP, \fI\%memcached_callback_t\fP\fI\ flag\fP, const void\fI\ *data\fP)
.UNINDENT
.INDENT 0.0
.TP
.B void * memcached_callback_get(memcached_st\fI\ *ptr\fP, \fI\%memcached_callback_t\fP\fI\ flag\fP, memcached_return_t\fI\ *error\fP)
.UNINDENT
.sp
Compile and link with \-lmemcached
.SH DESCRIPTION
.sp
libmemcached(3) can have callbacks set key execution points. These either
provide function calls at points in the code, or return pointers to
structures for particular usages.
.sp
\fI\%memcached_callback_get()\fP takes a callback flag and returns the
structure or function set by \fI\%memcached_callback_set()\fP.
.sp
\fI\%memcached_callback_set()\fP changes the function/structure assigned by a
callback flag. No connections are reset.
.sp
You can use \fI\%MEMCACHED_CALLBACK_USER_DATA\fP to provide custom context
if required for any of the callbacks.
.INDENT 0.0
.TP
.B MEMCACHED_CALLBACK_CLEANUP_FUNCTION
.UNINDENT
.sp
When \fBmemcached_delete()\fP is called this function will be excuted. At
the point of its execution all connections are closed.
.INDENT 0.0
.TP
.B MEMCACHED_CALLBACK_CLONE_FUNCTION
.UNINDENT
.sp
When \fBmemcached_delete()\fP is called this function will be excuted.
At the point of its execution all connections are closed.
.INDENT 0.0
.TP
.B MEMCACHED_CALLBACK_PREFIX_KEY
See \fI\%MEMCACHED_CALLBACK_NAMESPACE\fP
.UNINDENT
.INDENT 0.0
.TP
.B MEMCACHED_CALLBACK_NAMESPACE
.UNINDENT
.sp
You can set a value which will be used to create a domain for your keys.
The value specified here will be prefixed to each of your keys. The value can
not be greater then \fBMEMCACHED_PREFIX_KEY_MAX_SIZE\fP \- 1 and will
reduce \fBMEMCACHED_MAX_KEY\fP by the value of your key.
.sp
The prefix key is only applied to the primary key, not the master key.
\fBMEMCACHED_FAILURE\fP will be returned if no key is set. In the case of
a key which is too long, \fBMEMCACHED_BAD_KEY_PROVIDED\fP will be returned.
.sp
If you set a value with the value being NULL then the prefix key is disabled.
.INDENT 0.0
.TP
.B MEMCACHED_CALLBACK_USER_DATA
.UNINDENT
.sp
This allows you to store a pointer to a specifc piece of data. This can be
retrieved from inside of \fBmemcached_fetch_execute()\fP. Cloning a
\fBmemcached_st\fP will copy the pointer to the clone.
.INDENT 0.0
.TP
.B MEMCACHED_CALLBACK_MALLOC_FUNCTION
.UNINDENT
.sp
Deprecated since version <0.32: Use \fBmemcached_set_memory_allocators\fP instead.
.INDENT 0.0
.TP
.B MEMCACHED_CALLBACK_REALLOC_FUNCTION
.UNINDENT
.sp
Deprecated since version <0.32: Use \fBmemcached_set_memory_allocators\fP instead.
.INDENT 0.0
.TP
.B MEMCACHED_CALLBACK_FREE_FUNCTION
.UNINDENT
.sp
Deprecated since version <0.32: Use \fBmemcached_set_memory_allocators\fP instead.
.INDENT 0.0
.TP
.B MEMCACHED_CALLBACK_GET_FAILURE
.UNINDENT
.sp
This function implements the read through cache behavior. On failure of retrieval this callback will be called.
.sp
You are responsible for populating the result object provided. This result object will then be stored in the server and returned to the calling process.
.sp
You must clone the \fBmemcached_st\fP in order to
make use of it. The value will be stored only if you return
\fBMEMCACHED_SUCCESS\fP or \fBMEMCACHED_BUFFERED\fP. Returning
\fBMEMCACHED_BUFFERED\fP will cause the object to be buffered and not sent
immediatly (if this is the default behavior based on your connection setup
this will happen automatically).
.sp
The prototype for this is:
.INDENT 0.0
.TP
.B memcached_return_t (*memcached_trigger_key)(memcached_st *ptr, char *key, size_t key_length, memcached_result_st *result);
.UNINDENT
.INDENT 0.0
.TP
.B MEMCACHED_CALLBACK_DELETE_TRIGGER
.UNINDENT
.sp
This function implements a trigger upon successful deletion of a key. The memcached_st structure will need to be cloned in order to make use of it.
.sp
The prototype for this is:
.INDENT 0.0
.TP
.B typedef memcached_return_t (*memcached_trigger_delete_key)(memcached_st *ptr, char *key, size_t key_length);
.UNINDENT
.SH RETURN
.sp
\fI\%memcached_callback_get()\fP return the function or structure that was
provided. Upon error, nothing is set, null is returned, and the
\fBmemcached_return_t\fP argument is set to \fBMEMCACHED_FAILURE\fP.
.sp
\fI\%memcached_callback_set()\fP returns \fBMEMCACHED_SUCCESS\fP upon
successful setting, otherwise \fBMEMCACHED_FAILURE\fP on error.
.SH HOME
.sp
To find out more information please check:
\fI\%http://libmemcached.org/\fP
.SH AUTHOR
.sp
Brian Aker, <\fI\%brian@tangent.org\fP>
.SH SEE ALSO
.sp
\fImemcached(1)\fP \fIlibmemcached(3)\fP \fImemcached_strerror(3)\fP
.SH AUTHOR
Brian Aker
.SH COPYRIGHT
2011-2012, Brian Aker DataDifferential, http://datadifferential.com/
.\" Generated by docutils manpage writer.
.\" 
.


ChangeLog
1.0.11 Sun Sep 16 20:32:13 EDT 2012
* Removed custom version of memcached.
* Updated hardening rules.
* Fixed a case where the return error from a socket connection differred from that of a TCP/IP socket.

1.0.10 Sun Jul 29 21:50:15 PDT 2012
* --disable-assert has been removed from configure, and --enable-assert has been added in its place.
* Compiling fixes for Clang on OSX Mountain Lion.

1.0.9 Wed Jul  4 22:46:19 EDT 2012
* Faster close on socket.
* Instance allocation is now seperated from server interface. This should allow for a better preservation of ABI compliance from now on.
* Fix close on exec bug.
* Numerous other bug fixes.

1.0.8 Tue May 22 15:06:04 EDT 2012
* Added support for setting options via ENV variable LIBMEMCACHED
* Fix corner case on last used result.

1.0.7 Sat Apr 28 00:48:29 PDT 2012
* Add API call for exist calls.
* Update all license files to be BSD.

1.0.6 Sat Apr  7 18:26:49 PDT 2012
* Fixes for gcc 4.7, lp:961812
* Fix for restart issue that happens under testing.
* Fix for lp:962815.
* Support for transparent AES encryption.

1.0.5 Tue Mar 13 22:56:47 PDT 2012
* Fixes for OSX.
* Version is now parsed directly in the parser, which makes buffered operations now work with it..
* memstat has been extended so that it can be used to find the version of the server.
* Update documentation.
* Fixes for compile issues on Debian and Ubuntu

1.0.4 Thu Jan 26 22:33:54 PST 2012
* Fix for memcached_dump().
* Additional testing for memcached_stat_execute().

1.0.3 Sun Jan  8 18:11:36 PST 2012
* Increased size of sort buffer used during Ketama.
* Added support for new behavior to handle dead servers.
* Overall haul of UDP IO.
* Fixed C compile issue with memcached_exist()
* Numerous bug fixes.
* Clang support for OSX.
* All commands now using vector send support.

1.0.2 Mon Oct 24 08:00:48 PDT 2011
* Dropped libmemcached/memcached_util.h (undocumented header file)
* Added memcached_touch() and memcached_touch_by_key()
* UDP support restructured to toggle on a complete memcached_st structure.

0.53 Mon Sep 26 20:50:33 PDT 2011
  * Fix for FreeBSD/OpenBSD and -lm
  * Added memcached_exist()
  * Fix for memory when using config test.
  * CLI gained --quiet

0.52 Sun Sep 11 22:16:08 PDT 2011
  * Build fixes for Ubuntu/Suse.
  * Fixes for OSX Lion.
  * Bug fix for looping back through dns lookups under certain failures.
  * Fixes related to dead server failures.

0.51 Thu Jul 21 11:08:58 PDT 2011
  * memcached_callback_set() now takes its data argument as const
  * Update to tests.
  * Fix in parser for port number.

0.50 Mon Jun 20 10:36:57 PDT 2011
 * Updates to C++ interface 
 * Custom free allocators need to now check for value before calling free.
 * memcached_fetch_result() now uses the internal result when available (about 25 to 50% faster).
 * Fix for stats structure.
 * Updates to documentation.
 * memcached_fetch_result() now uses the internal result when available (about 25 to 50% faster).

0.49 Thu Apr 14 08:43:37 PDT 2011
  * Fix calls to auto methods so that if value is not passed in nothing bad happens.
  * New parser calls for generating memcached_st objects.
  * New error system.
  * New flow control for messages means faster get/set calls.
  * Added new documentation system.
  * A behavior change has been now made that if you specify a weight for any server, we enable the weight flag and do weight balancing.  
  * Added MEMCACHED_BEHAVIOR_REMOVE_FAILED_SERVERS to simplify the setting of AUTO REJECT for servers.

0.48 Tue Mar 15 23:05:18 PDT 2011
  * Fix memory leak in server parse.
  * Move test framework out to be its own library (easier to work with Gearman).


0.47 Wed Feb 23 18:46:48 PST 2011
  * Additional fixes for OpenBSD.
  * Bug fix 677609, 456080.
  * SIGPIPE fix for Linux send(). 
  * memcapable can now test ascii or binary based on flags.
  * Additional build fixes for SASL.


0.46 Mon Feb 14 10:28:01 PST 2011
  * Fixes a number of corner case bugs.
  * Fixes related to OpenBSD.
  * Better testing for protocol version.
  * Removes special case infinite wait on blocking setup.

0.45 Tue Feb  8 16:02:06 PST 2011
  * Add support for systemtap

0.44 Wed Sep 22 21:57:57 PDT 2010
  * Windows bug fixes.
  * Hudson port support in test harness.
  * Improved portability of test hanrness.
  * SASL fixes.

0.43 Wed Jul 28 16:29:47 PDT 2010
  * Added --args to memstat so that a greater range of values can be returned.
  * Prelimanary support for Windows.
  * memcached_stat_execute() merged. 

0.42 Tue Jul  6 12:29:50 PDT 2010
  * Mistake in libtool caused issue with library version

0.41 Wed Jun 30 13:15:19 PDT 2010
  * Added --file for memcat.
  * Added limemcached_ping() to libmemcached_util
  * Bugfix for some cases where connect would have issues with timeout.
  * Wrong value for errno given as error on an IO failure inside of poll.
  * Bug fix for issue where multiple interfaces with bad DNS were not being caught.

0.40 Thu Apr 22 19:01:25 PDT 2010
  * Placed retry logic in for busted resolvers
  * Add an ignore for SIGPIPE to solve OSX issues.
  * A couple of fixed for memcached_light server.
  * Updated to debug mode to track io_wait

0.39 Tue Apr  6 12:35:13 PDT 2010
  * Add support for prefix keys to binary protocol.
  * Remove the undocumented call memcached_server_remove().
  * The undocumented call memcached_server_by_key() now returns const.
  * memcached_server_error_reset() has been deprecated.
  * memcached_server_list() has been deprecated. Use memcached_server_cursor() to walk the servers found in a memcached_st() structure.
  * memcached_verbosity() can now be run concurrently with other operations.
  * SASL support.
  * Fixes memory leak found in EJECT HOSTS.

0.38 Wed Feb 10 09:40:33 PST 2010
  * C++ interface for libhashkit.
  * Modified memcached_set_memory_allocators() so that it requires a context pointer.
  * memcached_clone() now runs 5 times faster.
  * Functions used for callbacks are now given const memcached_st.

libmemcached/memcached_callback.c
/* LibMemcached
 * Copyright (C) 2006-2009 Brian Aker
 * All rights reserved.
 *
 * Use and distribution licensed under the BSD license.  See
 * the COPYING file in the parent directory for full text.
 *
 * Summary: Change any of the possible callbacks.
 *
 */

#include "common.h" 
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/tcp.h>

/* 
  These functions provide data and function callback support
*/

memcached_return_t memcached_callback_set(memcached_st *ptr, 
                                          memcached_callback_t flag, 
                                          void *data)
{
  switch (flag)
  {
  case MEMCACHED_CALLBACK_PREFIX_KEY:
    {
      char *key= (char *)data;

      if (key)
      {
        size_t key_length= strlen(key);

        if (memcached_key_test((const char **)&key, &key_length, 1) == MEMCACHED_BAD_KEY_PROVIDED)
        {
          return MEMCACHED_BAD_KEY_PROVIDED;
        }

        if ((key_length > MEMCACHED_PREFIX_KEY_MAX_SIZE -1)
            || (strcpy(ptr->prefix_key, key) == NULL))
        {
          ptr->prefix_key_length= 0;
          return MEMCACHED_BAD_KEY_PROVIDED;
        }
        else
        {
          ptr->prefix_key_length= key_length;
        }
      }
      else
      {
        memset(ptr->prefix_key, 0, MEMCACHED_PREFIX_KEY_MAX_SIZE);
        ptr->prefix_key_length= 0;
      }

      break;
    }
  case MEMCACHED_CALLBACK_USER_DATA:
    {
      ptr->user_data= data;
      break;
    }
  case MEMCACHED_CALLBACK_CLEANUP_FUNCTION:
    {
      memcached_cleanup_fn func= *(memcached_cleanup_fn *)&data;
      ptr->on_cleanup= func;
      break;
    }
  case MEMCACHED_CALLBACK_CLONE_FUNCTION:
    {
      memcached_clone_fn func= *(memcached_clone_fn *)&data;
      ptr->on_clone= func;
      break;
    }
#ifdef MEMCACHED_ENABLE_DEPRECATED
  case MEMCACHED_CALLBACK_MALLOC_FUNCTION:
    {
      memcached_malloc_function func= *(memcached_malloc_fn *)&data;
      ptr->call_malloc= func;
      break;
    }
  case MEMCACHED_CALLBACK_REALLOC_FUNCTION:
    {
      memcached_realloc_function func= *(memcached_realloc_fn *)&data;
      ptr->call_realloc= func;
      break;
    }
  case MEMCACHED_CALLBACK_FREE_FUNCTION:
    {
      memcached_free_function func= *(memcached_free_fn *)&data;
      ptr->call_free= func;
      break;
    }
#endif
  case MEMCACHED_CALLBACK_GET_FAILURE:
    {
      memcached_trigger_key_fn func= *(memcached_trigger_key_fn *)&data;
      ptr->get_key_failure= func;
      break;
    }
  case MEMCACHED_CALLBACK_DELETE_TRIGGER:
    {
      memcached_trigger_delete_key_fn func= *(memcached_trigger_delete_key_fn *)&data;
      ptr->delete_trigger= func;
      break;
    }
  default:
    return MEMCACHED_FAILURE;
  }

  return MEMCACHED_SUCCESS;
}

void *memcached_callback_get(memcached_st *ptr, 
                             memcached_callback_t flag,
                             memcached_return_t *error)
{
  memcached_return_t local_error;

  if (!error)
    error = &local_error;

  switch (flag)
  {
  case MEMCACHED_CALLBACK_PREFIX_KEY:
    {
      if (ptr->prefix_key[0] == 0)
      {
        *error= MEMCACHED_FAILURE;
        return NULL;
      }
      else
      {
        *error= MEMCACHED_SUCCESS;
        return (void *)ptr->prefix_key;
      }
    }
  case MEMCACHED_CALLBACK_USER_DATA:
    {
      *error= ptr->user_data ? MEMCACHED_SUCCESS : MEMCACHED_FAILURE;
      return (void *)ptr->user_data;
    }
  case MEMCACHED_CALLBACK_CLEANUP_FUNCTION:
    {
      *error= ptr->on_cleanup ? MEMCACHED_SUCCESS : MEMCACHED_FAILURE;
      return *(void **)&ptr->on_cleanup;
    }
  case MEMCACHED_CALLBACK_CLONE_FUNCTION:
    {
      *error= ptr->on_clone ? MEMCACHED_SUCCESS : MEMCACHED_FAILURE;
      return *(void **)&ptr->on_clone;
    }
#ifdef MEMCACHED_ENABLE_DEPRECATED
  case MEMCACHED_CALLBACK_MALLOC_FUNCTION:
    {
      *error= ptr->call_malloc ? MEMCACHED_SUCCESS : MEMCACHED_FAILURE;
      return *(void **)&ptr->call_malloc;
    }
  case MEMCACHED_CALLBACK_REALLOC_FUNCTION:
    {
      *error= ptr->call_realloc ? MEMCACHED_SUCCESS : MEMCACHED_FAILURE;
      return *(void **)&ptr->call_realloc;
    }
  case MEMCACHED_CALLBACK_FREE_FUNCTION:
    {
      *error= ptr->call_free ? MEMCACHED_SUCCESS : MEMCACHED_FAILURE;
      return *(void **)&ptr->call_free;
    }
#endif
  case MEMCACHED_CALLBACK_GET_FAILURE:
    {
      *error= ptr->get_key_failure ? MEMCACHED_SUCCESS : MEMCACHED_FAILURE;
      return *(void **)&ptr->get_key_failure;
    }
  case MEMCACHED_CALLBACK_DELETE_TRIGGER:
    {
      *error= ptr->delete_trigger ? MEMCACHED_SUCCESS : MEMCACHED_FAILURE;
      return *(void **)&ptr->delete_trigger;
    }
  default:
    WATCHPOINT_ASSERT(0);
    *error= MEMCACHED_FAILURE;
    return NULL;
  }
}
